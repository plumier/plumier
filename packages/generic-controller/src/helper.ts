import { RelationDecorator, RelationPropertyDecorator } from "@plumier/core"
import reflect, { Class, generic, GenericTypeDecorator } from "@plumier/reflect"


// --------------------------------------------------------------------- //
// ----------------------- ENTITY RELATION HELPER ---------------------- //
// --------------------------------------------------------------------- //

const genericControllerRegistry = new Map<Class, boolean>()

function updateGenericControllerRegistry(cls: Class) {
    genericControllerRegistry.set(cls, true)
}

function getGenericControllerRelation(controller: Class) {
    const meta = reflect(controller)
    const types = generic.getGenericTypeParameters(controller)
    const parentEntityType = types[0]
    const entityType = types[1]
    const oneToMany = meta.decorators.find((x: RelationPropertyDecorator): x is RelationPropertyDecorator => x.kind === "plumier-meta:relation-prop-name")
    const relation = oneToMany!.name
    return { parentEntityType, entityType, relation, inverseProperty: oneToMany!.inverseProperty }
}

function getGenericControllerInverseProperty(controller: Class) {
    const rel = getGenericControllerRelation(controller)
    return rel.inverseProperty
}

function getGenericControllerOneToOneRelations(type: Class) {
    const meta = reflect(type)
    const result = []
    for (const prop of meta.properties) {
        if (prop.decorators.find((x: RelationDecorator) => x.kind === "plumier-meta:relation") && prop.typeClassification !== "Array") {
            result.push(prop)
        }
    }
    return result
}


export {
    getGenericControllerOneToOneRelations, getGenericControllerInverseProperty,
    getGenericControllerRelation, updateGenericControllerRegistry, genericControllerRegistry
}