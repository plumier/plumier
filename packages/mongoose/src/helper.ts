import { authorize } from "@plumier/core"
import reflect, { Class, useCache } from "@plumier/reflect"

import { RefDecorator } from "./types"

function normalizeEntityNoCache(type: Class) {
    const meta = reflect(type)
    reflect.flush(type)
    for (const prop of meta.properties) {
        const ref = prop.decorators.find((x: RefDecorator) => x.name === "MongooseRef")
        if (ref && prop.typeClassification === "Array") {
            const decorators = [
                authorize.readonly(),
                authorize.writeonly()
            ]
            Reflect.decorate(decorators, type.prototype, prop.name, void 0)
        }
    }
    return { success: true }
}

const normalizeEntityCache = new Map<Class, any>()

export const normalizeEntity = useCache(normalizeEntityCache, normalizeEntityNoCache, x => x)