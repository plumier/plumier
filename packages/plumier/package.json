{
  "name": "plumier",
  "version": "1.0.0-beta.9",
  "description": "Delightful Node.js Rest Framework",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "keywords": [
    "MVC",
    "Koa",
    "Web API",
    "JSON Service",
    "TDD"
  ],
  "scripts": {
    "compile": "tsc -p tsconfig.build.json",
    "benchmark": "ts-node test/benchmark/run",
    "exe": "ts-node"
  },
  "author": "Ketut Sandiarsa",
  "license": "MIT",
  "dependencies": {
    "@koa/cors": "^3.0.0",
    "@plumier/core": "1.0.0-beta.9",
    "@types/faker": "^4.1.6",
    "@types/koa__cors": "^2.2.3",
    "faker": "^4.1.0",
    "koa": "^2.8.2",
    "koa-body": "^4.1.1"
  },
  "devDependencies": {
    "@plumier/jwt": "1.0.0-beta.9",
    "@plumier/mongoose": "1.0.0-beta.9",
    "@plumier/multipart": "1.0.0-beta.9",
    "@plumier/serve-static": "1.0.0-beta.9",
    "@plumier/social-login": "1.0.0-canary.31",
    "@types/body-parser": "^1.17.1",
    "@types/cors": "^2.8.6",
    "@types/express": "^4.17.1",
    "@types/express-jwt": "^0.0.42",
    "@types/fs-extra": "^8.0.0",
    "@types/joi": "^14.3.3",
    "@types/jsonwebtoken": "^8.3.4",
    "@types/koa-better-body": "^3.0.0",
    "@types/koa-router": "^7.0.42",
    "@types/rimraf": "^2.0.2",
    "@types/supertest": "^2.0.8",
    "autocannon": "^4.1.1",
    "benalu": "^2.0.0-beta-1",
    "body-parser": "^1.19.0",
    "cors": "^2.8.5",
    "express": "^4.17.1",
    "express-jwt": "^5.3.1",
    "fs-extra": "^8.1.0",
    "joi": "^14.3.1",
    "jsonwebtoken": "^8.5.1",
    "koa-better-body": "^3.1.2",
    "koa-router": "^7.4.0",
    "rimraf": "^3.0.0",
    "supertest": "^4.0.2",
    "ts-node": "^8.4.1",
    "typescript": "^3.6.3"
  },
  "bugs": {
    "url": "https://github.com/plumier/plumier/issues"
  },
  "repository": {
    "url": "https://github.com/plumier/plumier"
  },
  "homepage": "https://plumierjs.com",
  "publishConfig": {
    "access": "public"
  },
  "gitHead": "9d69b5621b7e2e777818e5e36c13ecba0c8d8204"
}
