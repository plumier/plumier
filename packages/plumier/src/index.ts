export {
    authorize,
    ActionResult,
    Application,
    bind,
    Configuration,
    Class,
    DependencyResolver,
    DefaultDependencyResolver,
    Facility,
    FileUploadInfo,
    FileParser,
    HeaderPart,
    HttpMethod,
    HttpStatus,
    HttpStatusError,
    Invocation,
    KoaMiddleware,
    middleware,
    Middleware,
    MiddlewareUtil,
    domain,
    PlumierApplication,
    PlumierConfiguration,
    RequestPart,
    route,
    RouteInfo,
    Authorizer, 
    AuthorizeMetadataInfo,
    CustomValidator,
    ValidatorFunction,
    ValidatorInfo,
    DefaultFacility,
    response,
    val, 
    AsyncValidatorResult
} from "@plumier/core"
export * from "./facility"

import { Plumier } from "./application"
export default Plumier