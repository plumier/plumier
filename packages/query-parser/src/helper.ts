import { AuthorizeDecorator, AuthorizerContext, authPolicy, AuthPolicy, EntityAuthPolicy, executeAuthorizer } from "@plumier/core"
import { Class, generic } from "@plumier/reflect"


const ParserAst = Symbol()

function getDecoratorType(controller: Class, expType: string | Class) {
    // extract generic type from controller if string type provided
    return typeof expType === "string" ? generic.getGenericType(controller, expType) as Class : expType
}

async function isAuthorized(decorators: AuthorizeDecorator[], ctx: AuthorizerContext): Promise<boolean> {
    const authPolicies = ctx.ctx.config.authPolicies.map(x => new x())
    const policies: { decorator: AuthorizeDecorator, isEntityPolicy: boolean }[] = []
    const staticPolicies = []
    // flatten decorator to make each decorator only have one policy
    // and add flag if its an entity policy
    // to make them easily filtered by its flag
    for (const decorator of decorators) {
        for (const policy of decorator.policies) {
            const isEntityPolicy = authPolicies.some(x => x.name === policy && x instanceof EntityAuthPolicy)
            const dec: AuthorizeDecorator = { ...decorator, policies: [policy] }
            if (!isEntityPolicy)
                staticPolicies.push(dec)
            policies.push({ decorator: dec, isEntityPolicy })
        }
    }
    if (policies.every(x => x.isEntityPolicy)) return false
    return executeAuthorizer(staticPolicies, ctx)
}

export { getDecoratorType, ParserAst, isAuthorized }