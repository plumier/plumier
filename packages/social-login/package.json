{
  "name": "@plumier/social-login",
  "version": "1.0.0-beta.10",
  "description": "OAuth module for Plumier",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "scripts": {
    "compile": "tsc -p tsconfig.build.json"
  },
  "author": "Ketut Sandiarsa",
  "license": "MIT",
  "dependencies": {
    "@plumier/core": "1.0.0-beta.10",
    "@types/axios": "0.14.0",
    "@types/debug": "^4.1.5",
    "axios": "0.19.0",
    "debug": "^4.1.1"
  },
  "bugs": {
    "url": "https://github.com/plumier/plumier/issues"
  },
  "repository": {
    "url": "https://github.com/plumier/plumier"
  },
  "homepage": "https://plumierjs.com",
  "publishConfig": {
    "access": "public"
  },
  "devDependencies": {
    "@types/dotenv": "^8.2.0",
    "dotenv": "^8.2.0"
  }
}
