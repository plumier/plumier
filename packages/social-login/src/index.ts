import { ActionResult, middleware, response } from "@plumier/core"

import { content } from "./callback-view"
import { DialogProvider, SocialAuthProvider, OAuthCallbackMiddleware, OAuthDialogEndPointMiddleware } from "./middleware"

export function oAuthCallback(option: SocialAuthProvider) {
    return middleware.use(new OAuthCallbackMiddleware(option))
}

export function oAuthDialogEndPoint(option:DialogProvider){
    return middleware.use(new OAuthDialogEndPointMiddleware(option))
}

export { GoogleProvider, GoogleLoginStatus, GoogleProfile, GoogleDialogProvider } from "./provider/google"
export { FacebookProvider, FacebookLoginStatus, FacebookProfile, FacebookDialogProvider } from "./provider/facebook"
export { GitHubProvider, GitHubLoginStatus, GitHubProfile, GitHubDialogProvider } from "./provider/github"
export { GitLabProvider, GitLabLoginStatus, GitLabProfile, GitLabDialogProvider } from "./provider/gitlab"
export { OAuthCallbackMiddleware, OAuthDialogEndPointMiddleware, SocialLoginStatus, SocialAuthProvider, DialogProvider } from "./middleware"

declare module "@plumier/core" {
    namespace response {
        function callbackView(message: any): ActionResult;
    }
}

response.callbackView = (message: any) => {
    return new ActionResult(content(message))
        .setHeader("Content-Type", "text/html")
}