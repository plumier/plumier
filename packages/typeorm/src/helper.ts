import { authorize, Class, entity } from "@plumier/core"
import reflect, { noop, useCache } from "@plumier/reflect"
import { getMetadataArgsStorage } from "typeorm"
import { parse } from "acorn"

function inverseSideParser(expr: ((t: any) => any)) {
    const node = parse(expr.toString(), { ecmaVersion: 2020 })
    return getMemberExpression(node)
}

function getContent(node: any): any {
    switch (node.type) {
        case "Program":
        case "BlockStatement":
            return node.body[node.body.length - 1]
        case "ArrowFunctionExpression":
            return node.body
        case "ExpressionStatement":
            return node.expression
        case "ReturnStatement":
            return node.argument
    }
}

function getMemberExpression(node: any): string {
    const content = getContent(node)
    if (content.type === "MemberExpression")
        return content.property.name
    else
        return getMemberExpression(content)
}


function normalizeEntityNoCache(type: Class) {
    const parent: Class = Object.getPrototypeOf(type)
    // loop through parent entities 
    if (!!parent.prototype) normalizeEntity(parent)
    const storage = getMetadataArgsStorage();
    const columns = storage.filterColumns(type)
    for (const col of columns) {
        Reflect.decorate([noop()], (col.target as Function).prototype, col.propertyName, void 0)
        if (col.options.primary)
            Reflect.decorate([entity.primaryId(), authorize.readonly()], (col.target as Function).prototype, col.propertyName, void 0)
    }
    const relations = storage.filterRelations(type)
    for (const col of relations) {
        const rawType: Class = (col as any).type()
        if (col.relationType === "many-to-many" || col.relationType === "one-to-many") {
            const inverseProperty = inverseSideParser(col.inverseSideProperty as any)
            const decorators = [
                reflect.type(x => [rawType]),
                entity.relation({ inverseProperty }),
                authorize.readonly(),
                authorize.writeonly()
            ]
            Reflect.decorate(decorators, (col.target as Function).prototype, col.propertyName, void 0)
        }
        else {
            Reflect.decorate([reflect.type(x => rawType), entity.relation()], (col.target as Function).prototype, col.propertyName, void 0)
        }
    }
}

const normalizeEntityCache = new Map<Class, any>()

export const normalizeEntity = useCache(normalizeEntityCache, normalizeEntityNoCache, x => x)