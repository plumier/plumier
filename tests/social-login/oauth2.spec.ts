import { CookieName, OAuthProviderOption, OAuthUser, redirectUri } from "@plumier/social-login"
import Axios from "axios"
import Csrf from "csrf"
import { Context } from "koa"
import { bind, Class, response, route } from "plumier"
import qs from "querystring"

import { fixture } from "../helper"
import { error, getLoginUrl, params, portRegex, runServer } from "./helper"
import { FakeOAuth2Facility, oAuthServer } from "./oauth2-server"


/**
 * Run 2 fake oauth servers and a single oauth consumer server.
 */
async function appStub(controller: Class, mazeOpt?: OAuthProviderOption, googolOpt?: OAuthProviderOption) {
    // run fake oauth servers 
    const mazeServer = await runServer(await oAuthServer())
    const googolServer = await runServer(await oAuthServer())
    // run oauth consumer server on another port
    const koa = await fixture(controller)
        .set(new FakeOAuth2Facility("MazeBook", mazeServer.origin, mazeOpt))
        .set(new FakeOAuth2Facility("Googol", googolServer.origin, googolOpt))
        .initialize()
    const consumer = await runServer(koa)
    return {
        origin: consumer.origin,
        close: () => {
            mazeServer.server.close();
            googolServer.server.close();
            consumer.server.close()
        }
    }
}


async function login(url: string): Promise<{ status: number, profile: any, authUrl: string, cookie: string[] }> {
    const loginResp = await Axios.get(url)
    const loginUrl = getLoginUrl(loginResp.data)
    const cookie: string = loginResp.headers["set-cookie"].join("; ")
    var profileResp: any
    try {
        profileResp = await Axios.post(loginUrl, undefined, { headers: { cookie } })
    } catch (e) {
        profileResp = e.response
    }
    return {
        status: profileResp.status,
        profile: profileResp.data,
        authUrl: loginUrl,
        cookie: loginResp.headers["set-cookie"]
    }
}

const opt = { clientSecret: "super secret", clientId: "1234" }

describe("OAuth 2.0", () => {
    class AuthController {
        @redirectUri()
        callback(@bind.oAuthUser() user: OAuthUser) {
            return user
        }
    }

    describe("Functionalities", () => {
        it("Should login properly", async () => {
            const servers = await appStub(AuthController, opt, opt)
            const mazeProfile = await login(`${servers.origin}/auth/mazebook/login`)
            const googolProfile = await login(`${servers.origin}/auth/googol/login`)
            expect({ ...mazeProfile, authUrl: "", cookie: "" }).toMatchSnapshot()
            expect({ ...googolProfile, authUrl: "", cookie: "" }).toMatchSnapshot()
            servers.close()
        })

        it("Should save provider on cookie", async () => {
            const servers = await appStub(AuthController, opt, opt)
            const { cookie } = await login(`${servers.origin}/auth/mazebook/login`)
            expect(cookie.find(x => x.match(CookieName.provider))).toMatchSnapshot()
            servers.close()
        })

        it("Should save CSRF secret on cookie and CSRF token on state parameter", async () => {
            const servers = await appStub(AuthController, opt, opt)
            const { cookie, authUrl } = await login(`${servers.origin}/auth/mazebook/login`)
            const tokenRegex = /^plum-oauth:csrf-secret=(.*?);.*/i
            const theCookie = cookie.find(x => x.match(CookieName.csrfSecret))!
            const secret = theCookie.match(tokenRegex)![1]
            const token = params(authUrl).state;
            const csrf = new Csrf()
            expect(csrf.verify(secret, token)).toBe(true)
            expect(theCookie.replace(secret, "")).toMatchSnapshot()
            servers.close()
        })

        it("Should attach proper redirect URI on login url", async () => {
            const servers = await appStub(AuthController, opt, opt)
            const { authUrl } = await login(`${servers.origin}/auth/mazebook/login`)
            const redirect = params(authUrl).redirect_uri.replace(portRegex, "")
            expect(redirect).toMatchSnapshot()
            servers.close()
        })

        it("Should use default environment variable", async () => {
            process.env.PLUM_MAZEBOOK_CLIENT_ID = "123"
            process.env.PLUM_MAZEBOOK_CLIENT_SECRET = "secret"
            process.env.PLUM_GOOGOL_CLIENT_ID = "123"
            process.env.PLUM_GOOGOL_CLIENT_SECRET = "secret"
            const servers = await appStub(AuthController)
            const { status } = await login(`${servers.origin}/auth/mazebook/login`)
            expect(status).toBe(200)
            servers.close()
            delete process.env.PLUM_MAZEBOOK_CLIENT_ID
            delete process.env.PLUM_MAZEBOOK_CLIENT_SECRET
            delete process.env.PLUM_GOOGOL_CLIENT_ID
            delete process.env.PLUM_GOOGOL_CLIENT_SECRET
        })

        it("Should able to change login url", async () => {
            const servers = await appStub(AuthController, {
                clientSecret: "super secret",
                clientId: "1234",
                loginEndPoint: "/auth/mazebook/sign-in",
            }, opt)
            const { status } = await login(`${servers.origin}/auth/mazebook/sign-in`)
            expect(status).toBe(200)
            servers.close()
        })

        it("Should able add extra profile params", async () => {
            const servers = await appStub(AuthController, {
                clientSecret: "super secret",
                clientId: "1234",
                profileParams: { lorem: "ipsum", dolor: "sit" }
            }, opt)
            const { authUrl, cookie, ...profile } = await login(`${servers.origin}/auth/mazebook/login`)
            expect(profile).toMatchSnapshot()
            servers.close()
        })

        it("Should able to distinguish global callback and specific callback", async () => {
            class AuthController {
                @redirectUri()
                callback(@bind.oAuthUser() user: OAuthUser) {
                    return user
                }

                @redirectUri("MazeBook" as any)
                fake(@bind.oAuthUser() user: OAuthUser) {
                    return { lorem: "ipsum" }
                }
            }
            const servers = await appStub(AuthController, opt, opt)
            const { authUrl, cookie, ...profile } = await login(`${servers.origin}/auth/mazebook/login`)
            expect(profile).toMatchSnapshot()
            servers.close()
        })

        it("Should able to send postMessage UI", async () => {
            class AuthController {
                @redirectUri()
                callback(@bind.oAuthUser() user: OAuthUser) {
                    return response.postMessage(user)
                }
            }
            const servers = await appStub(AuthController, opt, opt)
            const profile = await login(`${servers.origin}/auth/mazebook/login`)
            expect(profile.profile.replace(/var message = '(.*)';/, "")).toMatchSnapshot()
            servers.close()
        })

        it("Should able to set postMessage origin", async () => {
            class AuthController {
                @redirectUri()
                callback(@bind.oAuthUser() user: OAuthUser, @bind.ctx() ctx:Context) {
                    return response.postMessage(user, ctx.origin)
                }
            }
            const servers = await appStub(AuthController, opt, opt)
            const profile = await login(`${servers.origin}/auth/mazebook/login`)
            expect(profile.profile.replace(/var message = '(.*)';/, "").replace(portRegex, "")).toMatchSnapshot()
            servers.close()
        })
    })

    describe("Durability", () => {
        it("Should throw error when no Client ID provided", async () => {
            const plum = fixture(AuthController)
                .set(new FakeOAuth2Facility("MazeBook", "http://mazebook.com", opt))
                .set(new FakeOAuth2Facility("Googol", "http://googol.com", { clientSecret: "lorem" }))
            const fn = await error(plum.initialize())
            expect(fn.mock.calls).toMatchSnapshot()
        })

        it("Should throw error when no Client SECRET provided", async () => {
            const plum = fixture(AuthController)
                .set(new FakeOAuth2Facility("MazeBook", "http://mazebook.com", { clientId: "123" }))
                .set(new FakeOAuth2Facility("Googol", "http://googol.com", opt))
            const fn = await error(plum.initialize())
            expect(fn.mock.calls).toMatchSnapshot()
        })

        it("Should throw error when no redirect uri provided", async () => {
            class AnimalController {
                @redirectUri("Googol" as any)
                callback(@bind.oAuthUser() user: OAuthUser) {
                    return user
                }
            }
            const plum = fixture(AnimalController)
                .set(new FakeOAuth2Facility("Googol", "http://googol.com", opt))
                .set(new FakeOAuth2Facility("MazeBook", "http://mazebook.com", { clientId: "123" }))
            const fn = await error(plum.initialize())
            expect(fn.mock.calls).toMatchSnapshot()
        })

        it("Should throw error when provided redirect uri with parameter", async () => {
            class AnimalController {
                @redirectUri()
                @route.get("/callback/:id")
                callback(id: string, @bind.oAuthUser() user: OAuthUser) {
                    return user
                }
            }
            const plum = fixture(AnimalController)
                .set(new FakeOAuth2Facility("Googol", "http://googol.com", opt))
                .set(new FakeOAuth2Facility("MazeBook", "http://mazebook.com", { clientId: "123" }))
            const fn = await error(plum.initialize())
            expect(fn.mock.calls).toMatchSnapshot()
        })

        it("Should throw error when provided non GET redirect uri", async () => {
            class AnimalController {
                @redirectUri()
                @route.post()
                callback(@bind.oAuthUser() user: OAuthUser) {
                    return user
                }
            }
            const plum = fixture(AnimalController)
                .set(new FakeOAuth2Facility("Googol", "http://googol.com", opt))
                .set(new FakeOAuth2Facility("MazeBook", "http://mazebook.com", { clientId: "123" }))
            const fn = await error(plum.initialize())
            expect(fn.mock.calls).toMatchSnapshot()
        })

        it("Should check for invalid CSRF token", async () => {
            const servers = await appStub(AuthController, opt, opt)
            const loginResp = await Axios.get(`${servers.origin}/auth/mazebook/login`)
            const loginUrl = /const REDIRECT = '(.*)'/.exec(loginResp.data)![1]
            const pars = params(loginUrl)
            delete pars.state
            const newLoginUrl = loginUrl.split("?")![0] + "?" + qs.stringify(pars)
            const cookie: string = loginResp.headers["set-cookie"].join("; ")
            const fn = jest.fn()
            try {
                await Axios.post(newLoginUrl, undefined, { headers: { cookie } })
            } catch (e) {
                fn(e.response)
            }
            expect(fn.mock.calls[0][0].status).toBe(400)
            servers.close()
        })

        it("Should show proper error when internal OAuth server error occur", async () => {
            class AuthController {
                @redirectUri()
                @route.get("/auth/invalid-redirect")
                callback(@bind.oAuthUser() user: OAuthUser) {
                    return user
                }
            }
            const servers = await appStub(AuthController, opt, opt)
            const { authUrl, cookie, ...profile } = await login(`${servers.origin}/auth/mazebook/login`)
            expect(profile).toMatchSnapshot()
            servers.close()
        })
    })

})