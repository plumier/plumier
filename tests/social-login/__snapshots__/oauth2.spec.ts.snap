// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OAuth 2.0 Durability Should show proper error when internal OAuth server error occur 1`] = `
Object {
  "profile": Object {
    "message": "Bad Request",
    "status": 400,
  },
  "status": 400,
}
`;

exports[`OAuth 2.0 Durability Should throw error when no Client ID provided 1`] = `
Array [
  Array [
    "OAuth: Client id for Googol not provided",
  ],
]
`;

exports[`OAuth 2.0 Durability Should throw error when no Client SECRET provided 1`] = `
Array [
  Array [
    "OAuth: Client secret for MazeBook not provided",
  ],
]
`;

exports[`OAuth 2.0 Durability Should throw error when no redirect uri provided 1`] = `
Array [
  Array [
    "OAuth: No MazeBook redirect uri handler found",
  ],
]
`;

exports[`OAuth 2.0 Durability Should throw error when provided non GET redirect uri 1`] = `
Array [
  Array [
    "OAuth: Redirect uri must have GET http method on AnimalController.callback",
  ],
]
`;

exports[`OAuth 2.0 Durability Should throw error when provided redirect uri with parameter 1`] = `
Array [
  Array [
    "OAuth: Parameterized route is not supported on Googol callback uri",
  ],
]
`;

exports[`OAuth 2.0 Functionalities Should able add extra profile params 1`] = `
Object {
  "profile": Object {
    "email": "Josianne_Bosco@gmail.com",
    "family_name": "Erdman",
    "firstName": "I",
    "given_name": "Greta",
    "id": "54f483f9-9b50-4c15-b53f-46ece9e3e5f8",
    "lastName": "Ketut Manuel Auer",
    "locale": "en",
    "name": "I Ketut Manuel Auer",
    "picture": "http://lorempixel.com/640/480",
    "provider": "MazeBook",
    "query": Object {
      "dolor": "sit",
      "lorem": "ipsum",
    },
  },
  "status": 200,
}
`;

exports[`OAuth 2.0 Functionalities Should able to distinguish global callback and specific callback 1`] = `
Object {
  "profile": Object {
    "lorem": "ipsum",
  },
  "status": 200,
}
`;

exports[`OAuth 2.0 Functionalities Should able to send postMessage UI 1`] = `
"
    <!DOCTYPE html>
    <html>
        <title></title>
        <body>
            <div class=\\"container\\"></div>
            <script type=\\"text/javascript\\">
                
                var origin ;
                (function(){
                    window.onbeforeunload = function () {
                        window.opener.postMessage({ status: \\"Close\\" }, origin || window.location.origin);
                    };
                    window.opener.postMessage(JSON.parse(message), origin || window.location.origin);
                })()
            </script>
        </body>
    </html>
    "
`;

exports[`OAuth 2.0 Functionalities Should able to set postMessage origin 1`] = `
"
    <!DOCTYPE html>
    <html>
        <title></title>
        <body>
            <div class=\\"container\\"></div>
            <script type=\\"text/javascript\\">
                
                var origin = 'http://localhost';
                (function(){
                    window.onbeforeunload = function () {
                        window.opener.postMessage({ status: \\"Close\\" }, origin || window.location.origin);
                    };
                    window.opener.postMessage(JSON.parse(message), origin || window.location.origin);
                })()
            </script>
        </body>
    </html>
    "
`;

exports[`OAuth 2.0 Functionalities Should attach proper redirect URI on login url 1`] = `"http://localhost/auth/callback"`;

exports[`OAuth 2.0 Functionalities Should login properly 1`] = `
Object {
  "authUrl": "",
  "cookie": "",
  "profile": Object {
    "email": "Josianne_Bosco@gmail.com",
    "family_name": "Erdman",
    "firstName": "I",
    "given_name": "Greta",
    "id": "54f483f9-9b50-4c15-b53f-46ece9e3e5f8",
    "lastName": "Ketut Manuel Auer",
    "locale": "en",
    "name": "I Ketut Manuel Auer",
    "picture": "http://lorempixel.com/640/480",
    "provider": "MazeBook",
    "query": Object {},
  },
  "status": 200,
}
`;

exports[`OAuth 2.0 Functionalities Should login properly 2`] = `
Object {
  "authUrl": "",
  "cookie": "",
  "profile": Object {
    "email": "Josianne_Bosco@gmail.com",
    "family_name": "Erdman",
    "firstName": "I",
    "given_name": "Greta",
    "id": "54f483f9-9b50-4c15-b53f-46ece9e3e5f8",
    "lastName": "Ketut Manuel Auer",
    "locale": "en",
    "name": "I Ketut Manuel Auer",
    "picture": "http://lorempixel.com/640/480",
    "provider": "Googol",
    "query": Object {},
  },
  "status": 200,
}
`;

exports[`OAuth 2.0 Functionalities Should save CSRF secret on cookie and CSRF token on state parameter 1`] = `"plum-oauth:csrf-secret=; path=/; httponly"`;

exports[`OAuth 2.0 Functionalities Should save provider on cookie 1`] = `"plum-oauth:provider=MazeBook; path=/; httponly"`;
